src/ucp/api/ucp.h

```diff
@@ -1878,6 +1981,85 @@ ucs_status_ptr_t ucp_tag_send_sync_nb(ucp_ep_h ep, const void *buffer, size_t co
+ *                          receive operation is completed and the data is ready
+ *                          in the receive @a buffer. It is important to note
+ *                          that the call-back is only invoked in a case when
+ *                          the operation cannot be completed immediately.
+ * @param [in]     flags    Reserved for future use.
+ *
+ * @return UCS_OK               - The receive operation was completed
+ *                                immediately.
+ * @return UCS_PTR_IS_ERR(_ptr) - The receive operation failed.
+ * @return otherwise            - Operation was scheduled for receive. The
+ *                                request handle is returned to the application
+ *                                in order to track progress of the operation.
+ *                                The application is responsible to released the
+ *                                handle using @ref ucp_request_free
+ *                                "ucp_request_free()" routine.
+ */
+ucs_status_ptr_t ucp_stream_recv_nb(ucp_ep_h ep, void *buffer, size_t *count,
+                                    ucp_datatype_t datatype,
+                                    ucp_stream_recv_callback_t cb,
+                                    unsigned flags);
+
+
+/**
+ * @ingroup UCP_COMM
+ * @brief Non-blocking stream receive data.
+ *
+ * This routine receives a raw data on endpoint @ref ep if available that is
+ * described by UCS_STATUS_PTR(_ptr) in return value as a pointer to the data
+ * and @a length. The routine is a non-blocking and therefore returns
+ * immediately.
+ *
+ * @param [in]   ep               UCP endpoint that is used for the receive
+ *                                operation.
+ * @param [out]  length           Length of received data.
+ *
+ * @return UCS_OK               - No received data available on the @a ep.
+ * @return UCS_PTR_IS_ERR(_ptr) - the receive operation failed and
+ *                                UCS_PTR_STATUS(_ptr) indicates an error.
+ * @return otherwise            - The pointer to the data UCS_STATUS_PTR(_ptr)
+ *                                is returned to the application in order to
+ *                                capture internal UCP resources until received
+ *                                data is processed. When data is processed,
+ *                                application is responsible to release the
+ *                                handle using @ref ucp_stream_data_release
+ *                                "ucp_stream_data_release()" routine.
+ *
+ * @note This function returns packed data (equivalent to ucp_dt_make_contig(1)).
+ * @note This function returns a pointer to a UCP internal buffer, whereas
+ *       @ref ucp_stream_recv_nb places the data into a user-provided buffer.
+ *       In some cases, retrieving the internal buffer can be more optimal,
+         for example by processing the incoming data in-place and thus avoiding
+ *       extra memory copy operations.
```

Comment:
Suggest the following rewording:
This function returns a pointer to a UCP-supplied buffer, whereas ucp_stream_recv_nb places the data into a user-provided buffer. In some cases, receiving data directly into a UCP-supplied buffer can be more optimal, for example by processing the incoming data in-place and thus avoiding extra memory copy operations.